keps/sig-api-machinery/0032-20181026-flexible-inflight-request-limit.md

```diff
@@ -0,0 +1,153 @@
+---
+kep-number: 32
+title: Flexible inflight request limit
+authors:
+  - "@mshaverdo"
+owning-sig: sig-api-machinery
+reviewers:
+  - @wojtek-t
+  - @lavalamp
+  - @yliaog
+  - @jpbetz
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2018-10-26
+status: provisional
+---
+
+# Flexible inflight request limit
+
+## Table of Contents
+
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+* [Proposal](#proposal)
+    * [Algorithm to accept/reject readonly request](#algorithm-to-accept-reject-readonly-request)
+    * [How to determinate request weight](#how-to-determinate-request-weight)
+* [Discussionable questions](#discussionable-questions)
+    * [How to treat namespace-scoped lists?](#how-to-treat-namespace-scoped-lists)
+    * [Should we  make users connect to all apiservers to use their entire allotted qps?](#should-we-make-users-connect-to-all-apiservers)
+
+## Summary <a name="summary"></a>
```

Comment:
I think this KEP isn't quite ready for line-by-line comments yet. Here's some overall thoughts:

* This is an important thing for us to do relatively soon, but it's vitally important to get it right. So forgive us if we nitpick this a lot.
* I think there's N separable things we should divide this KEP into. At least clearly separate sections if not separate KEPs.

1. API call cost computation
  a. I'm still thinking about abstract tokens as the unit of account; if we can express this as a weighted combination of CPU & RAM then it might be clear how much quota apiserver has, and easier to reason about than completely abstract units.
  b. I'm very nervous about approaches that require keeping counts of objects per collection / namespace. I'd like to not do that, even if it makes the token computation less accurate. Pagination may help a lot with this.
2. How does the sysadmin set quota for users/groups/service accounts/namespaces/???
  a. This is an API design question, but also a statement that anything expressible is cheap to enforce.
3. How does apiserver enforce those limits.
  a. In particular, do we treat 3x apiservers as meaning we have 3x the quota to hand out?
  b. I don't think "just using a load balancer" is an answer because it's difficult for a user to control which apiserver they get connected to and therefore difficult for them to use their quota on all apiservers.
  c. OTOH if apiservers pool the quota (so that it doesn't matter which one the user connects to) then apiserver has a major coordination problem.
  d. Whatever we do has to be cheap enough computationally that we can reject, say, 10-100x as many requests as we serve. E.g., if we have quota to serve 100QPS, we should be able to reject an additional 1000-100000 QPS without impacting service. (This KEP should figure out a better estimate of this value, I made this one up.)

I think I favor a 2 tier approach, with tier 1 quota being guaranteed, and any tier 1 quota that goes unused is distributed ("re-sold") to tier 2. This allows promising certain components (scheduler, kubelet) quota while allowing others to burst when the cluster isn't busy. We obviously don't need to do that in the first take but we should keep it in mind.

Reply:
My 2cents about that:

> API call cost computation

I wouldn't try to convert the tokens to "cpu+ram combination" (at least not at the level of the proposal). Using some abstract concepts (even if they aren't very accurate) can be easier to understand. That's why I was suggesting that e.g. cost of a list can be considered as proportional to the number of items to be processed.
That said, I completely agree that keeping number of objects in memory per whatever is not what we want. One thing that I mentioned somwhere is that we may consider list to have the same cost no matter what exactly it is listing (it's obviously not very accurate, but maybe it's good enough first step)?
Re pagination - that helps if you specify limit, but you can't force people (at least now) to use that, right?

> How does apiserver enforce those limits.

Yes - this is definitely an important aspect. I was conceptually treating it as a separate feature (thus separate KEP), but probably you're right it should be mentioned here (at least some consequences of it).

3x apiservers won't mean 3x quota in all cases. As an example, etcd has a limitted write throughput, so no matter how many more apiserver you will throw, you won't get more quota than X.
I completely agree with the point that computing it has to be cheap. In particular, it definitely cannot trigger any read/write to the storage. It's probably debatable whether it can trigger some external call (I would say that probably not synchronously, but that's something that should be discussed)

> 2 tier approach

I think this would also need to be discussed deeper here. I agree this has advantages, but it has also disadvantages (e.g. do really have enough capacity to server provide tier1 quota for all system components if they would like to do everything at fully speed). It is also harder to reason about a bit in my opinion (I mean during execution< I guess I mean debugging more than reasoning). It's possible that it's a better model, but we definitely need a list of pros and cons to make a conscious decision.